### DISCUSSION TITLE
Pool Creation Workflows

### DISCUSSION THREAD

**Discussion by dszidi** [truenas/webui#7298] [Design Previews]
## Automated and Manual Disk Selection

It's been a long time coming but we're finally in a place where we can tackle the pool creation process.
The main problems we're trying to resolve with these designs:

1. Make this more scalable for larger deployments
2. Encourage best practices
3. Improve layout by making better use of screen real estate

With TrueNAS we've always walked a fine line between encouraging best practices and allowing users the freedom to do exactly what they need. We feel the way to do this with pool creation is to have separate automated and manual workflows.

## Automated Disk Selection

<img width="2168" alt="image" src="https://user-images.githubusercontent.com/29865743/201410173-b30a2802-01f6-4018-b9df-82ecff5149ad.png">

The first screen shows a wizard that breaks up the process into the following steps.

1. Specify name and encryption options
2. Create Data VDEVs
3. Create Log VDEVs
4. Create Space VDEVs
5. Create Cache VDEVs
6. Create Metadata VDEVs
7. Review/Submit VDEVs

In this scenario, instead of hand selecting disks, users will be describing what their VDEVs should look like and the UI will select the disks for them. The fields we're exposing in the form include the following:

- VDEV layout (MIRROR, RAIDZ1, RAIDZ2 etc)
- VDEV width (number of disks per VDEV)
- Disk size/type
- Number of VDEVs to create. 

### Context Cards

To provide context for the user, the UI will also provide a couple of cards on the side showing a disk inventory and configuration preview. These cards will update as the user makes their selections. This way they can do things like make sure they leave some disks available for other VDEV types as they progress through the wizard.

<img width="1180" alt="image" src="https://user-images.githubusercontent.com/29865743/201410338-22e97389-f1af-412a-adc0-1b8b9d178d84.png">

### Advanced Options

Having symmetrical VDEVs is best practice in the vast majority of cases. Especially on large systems with many disks across multiple shelves. For this reason this will be the default approach. We will try to accommodate certain edge cases via the Advanced Options section. If the user's needs are not covered by these then there is the button that will take the user into the manual selection workflow

<img width="937" alt="image" src="https://user-images.githubusercontent.com/29865743/200861908-a11f69a8-3924-4067-9d7f-dbabed6276a0.png">

## Manual Disk Selection

If the user has opted to manually select disks they will be presented with a large modal. There is a panel on the left where available disk devices are listed and the section on the right will show VDEVs that have been created. Users simply drag the device from the left into the area on the right

<img width="2168" alt="image" src="https://user-images.githubusercontent.com/29865743/200863772-431f9db2-d11e-4ea2-a6a5-5e6959f9cb2e.png">

>
>
> NOTE: The manual disk selection workflow is also what users will get when they want to add a vdev to an existing pool via the Storage/Devices page's "Add VDEV" button
>
>

### Disks Panel

Disks will now be represented as an icon to help save space. We've also included some filtering options as well as grouping disks by enclosure if enclosure support is available.

<img width="1321" alt="image" src="https://user-images.githubusercontent.com/29865743/200869349-e3dabda6-a30c-4f5c-9b36-48ea2aefd0a5.png">

>
>
> New custom icon representing enclosures helps save space.
>
>
> <img width="160" alt="image" src="https://user-images.githubusercontent.com/29865743/200871478-64f2c3e8-da4c-4c0c-b36f-cee397985b49.png">
>
>

### VDEVS

Thanks to the new icon based representation of disk devices, we can now show all the devices in a row. In addition to saving us vertical space, we can also group the disks by enclosure if enclosure support is available. Instead of using a string label, we're using our new enclosure icon. This comes in handy when users want maximum enclosure dispersal and every disk is on a separate enclosure. 

<img width="1597" alt="image" src="https://user-images.githubusercontent.com/29865743/200869626-987b9fc8-b276-41c2-a560-4d3fd311fa29.png">

<img width="1597" alt="image" src="https://user-images.githubusercontent.com/29865743/200869743-c6b76d2c-1b97-4be5-b3ad-3608debe58ab.png">

### Post Manual Selection

After the user has submitted their manual selections, they will be returned to the wizard. However once the user opts for manual disk selection, they must stay with it or start over. This is reflected in the areas that previously held the VDEV related fields and advanced options.

<img width="1405" alt="image" src="https://user-images.githubusercontent.com/29865743/200863991-578cd198-7fcd-4468-a6da-93e05abb02b0.png">

**Comment by Alexey-V-Gubin** [comment]
The "Configuration Preview" card lists
`Spare 12x RAIDZ2 | 12x 12TB`
What does RAIDZ2 mean when applied to spares?

Then, vdev disk list representation is horizontally oriented. What happens with very wide vdev, like 60 disks. I am thinking about dRAID, even if it is not supported now.

**Reply by dszidi** [reply]
Wrt the spare in the context cards, that was an oversight (repeat function in the design software). I'll correct it.

The horizontal vdev disk list will wrap into a grid layout. It might not be as neat and tidy as the mockups posted above, but it will still be a big improvement over the current table based approach. 

Below is another mockup showing what a vdev with 72 disks might look like. To be fair, we're not anticipating many people doing this via manual disk selection. For this use case, we'd expect users to take advantage of the automated workflow, as manually selecting 72 disks would be incredibly tedious.

<img width="2168" alt="image" src="https://user-images.githubusercontent.com/29865743/201408375-b5805e5e-7829-493a-a660-988018791bb7.png">

**Comment by w01230** [comment]
Seems awesome, looking forward in the futrue release.

**Comment by souporman** [comment]
> Specify name and encryption options
> Create Data VDEVs
> Create Log VDEVs
> Create Space VDEVs
> Create Cache VDEVs
> Create Metadata VDEVs
> Review/Submit VDEVs

Where would the dedupe tables go in this scenario?
